src/olympia/addons/decorators.py

```diff
@@ -40,9 +41,10 @@ def wrapper(request, addon_id=None, *args, **kw):
                     url += '?' + request.GET.urlencode()
                 return http.HttpResponsePermanentRedirect(url)
 
-        # If the addon is unlisted it needs either an owner/viewer/dev/support,
-        # or an unlisted addon reviewer.
-        if not (addon.has_listed_versions() or
+        # If the addon has no listed versions it needs either an author
+        # (owner/viewer/dev/support) or an unlisted addon reviewer.
+        if not (addon.has_listed_versions(
+            include_deleted=include_deleted_when_checking_versions) or
                 owner_or_unlisted_reviewer(request, addon)):
```

Comment:
I wonder if we can extract those function calls into separate variables above to make it easier to read:

```python
has_listed_version = addon.has_listed_versions(...)
is_owner_or_unlisted_reviewer = owner_or_unlisted_reviewer(...)

if not has_listed_version or is_owner_or_unlisted_reviewer:
    # ....
```

Reply:
I've extracted the first one into a separate variable. The second one would make a database query to load the user groups... It's likely the view using that decorator would need to do that at some point, and it's cached into an instance property, but still I'm not sure we want to do that just yet...

Reply:
That works for me, it's already much more readable. Thanks!